datasets/openslr/openslr.py

```diff
@@ -0,0 +1,216 @@
+
+
+class OpenSlr(datasets.GeneratorBasedBuilder):
+
+    BUILDER_CONFIGS = [
+        OpenSlrConfig(
+            name=resource_id,
+            language=_RESOURCES[resource_id]["Language"],
+            long_name=_RESOURCES[resource_id]["LongName"],
+            category=_RESOURCES[resource_id]["Category"],
+            summary=_RESOURCES[resource_id]["Summary"],
+            files=_RESOURCES[resource_id]["Files"],
+            index_files=_RESOURCES[resource_id]["IndexFiles"],
+            data_dirs=_RESOURCES[resource_id]["DataDirs"],
+        )
+        for resource_id in _RESOURCES.keys()
+    ]
+
+    def _info(self):
+        features = datasets.Features(
+            {
+                "path": datasets.Value("string"),
+                "sentence": datasets.Value("string"),
+            }
+        )
+
+        return datasets.DatasetInfo(
+            description=_DESCRIPTION,
+            features=features,
+            supervised_keys=None,
+            homepage=_HOMEPAGE,
+            license=_LICENSE,
+            citation=_CITATION,
+        )
+
+    def _split_generators(self, dl_manager):
+        """Returns SplitGenerators."""
+        resource_number = self.config.name.replace("SLR", "")
+        urls = [f"{_DATA_URL.format(resource_number)}/{file}" for file in self.config.files]
+        dl_paths = dl_manager.download_and_extract(urls)
+        abs_path_to_indexs = [os.path.join(path, f"{self.config.index_files[i]}") for i, path in enumerate(dl_paths)]
+        abs_path_to_datas = [os.path.join(path, f"{self.config.data_dirs[i]}") for i, path in enumerate(dl_paths)]
+
+        return [
+            datasets.SplitGenerator(
+                name=datasets.Split.TRAIN,
+                gen_kwargs={
+                    "path_to_indexs": abs_path_to_indexs,
+                    "path_to_datas": abs_path_to_datas,
+                },
+            ),
+        ]
+
+    def _generate_examples(self, path_to_indexs, path_to_datas):
+        """ Yields examples. """
+        data_fields = list(self._info().features.keys())
+
+        counter = -1
+        for i, path in enumerate(path_to_indexs):
+            with open(path, encoding="utf-8") as f:
+                lines = f.readlines()
+                for id_, line in enumerate(lines):
+                    # Following regexs are needed to normalise the lines, since the datasets
+                    # are not always consistent and have bugs:
+                    line = re.sub(r"\t[^\t]*\t", "\t", line.strip())
+                    field_values = re.split(r"\t\t?", line)
+                    # set absolute path for audio file
+                    if len(field_values) != 2:
+                        continue
+                    field_values[0] = os.path.join(path_to_datas[i], f"{field_values[0]}.wav")
+                    counter += 1
+                    yield counter, {key: value for key, value in zip(data_fields, field_values)}
```

Comment:
Maybe separating the path from the sentence would make this code easier to read ?

```suggestion
                    field_values = re.split(r"\t\t?", line)
                    if len(field_values) != 2:
                        continue
                    filename, sentence = field_values
                    # set absolute path for audio file
                    path = os.path.join(path_to_datas[i], f"{filename}.wav")
                    counter += 1
                    yield counter, {"path": path, "sentence": sentence}
```

Reply:
Thanks for the suggestion.